### PR TITLE
jni: added wrappers for global and local references

### PR DESCRIPTION
**Pull request by alex-koren**

### CODE REVIEW

File: src/include/baselib/jni/JniEnvironment.h [hunk 1]
```diff
@@ -136,6 +137,61 @@ namespace bl
...
+            }
+
+            void deleteGlobalRef( SAA_in const jobject object) const NOEXCEPT
+            {
+                m_jniEnv -> DeleteGlobalRef( object );
+            }
+
+            LocalReference< jclass > findJavaClass( SAA_in const std::string& javaClassName ) const
+            {
+                const auto javaClass = m_jniEnv -> FindClass( javaClassName.c_str() );
```
Comment (lazar-ivanov) [inline src/include/baselib/jni/JniEnvironment.h]: you need to immediately wrap this into local reference wrapper to make sure the code is always exception safe (no matter how it changes in the future)

File: src/include/baselib/jni/JniEnvironment.h [hunk 2]
```diff
@@ -136,6 +137,61 @@ namespace bl
...
+            }
+
+            void deleteLocalRef( SAA_in const jobject object) const NOEXCEPT
+            {
+                m_jniEnv -> DeleteLocalRef( object );
+            }
+
+            jobject newGlobalRef( SAA_in const jobject object ) const
+            {
+                const auto globalRef = m_jniEnv -> NewGlobalRef( object );
```
Comment (lazar-ivanov) [inline src/include/baselib/jni/JniEnvironment.h]: you need to immediately wrap this into global reference wrapper to make sure the code is always exception safe (no matter how it changes in the future)

File: src/include/baselib/jni/JniEnvironment.h [hunk 3]
```diff
@@ -136,6 +137,61 @@ namespace bl
                 return m_jniEnv -> GetVersion();
             }
 
+            jobjectRefType getObjectRefType( SAA_in const jobject object ) const NOEXCEPT
+            {
+                return m_jniEnv -> GetObjectRefType( object );
+            }
+
+            void deleteLocalRef( SAA_in const jobject object) const NOEXCEPT
```
Comment (lazar-ivanov) [inline src/include/baselib/jni/JniEnvironment.h]: space before )

File: src/include/baselib/jni/JniEnvironment.h [hunk 4]
```diff
@@ -136,6 +137,61 @@ namespace bl
                 return m_jniEnv -> GetVersion();
             }
 
+            jobjectRefType getObjectRefType( SAA_in const jobject object ) const NOEXCEPT
+            {
+                return m_jniEnv -> GetObjectRefType( object );
```
Comment (lazar-ivanov) [inline src/include/baselib/jni/JniEnvironment.h]: error handling needed?

File: src/include/baselib/jni/JniEnvironment.h [hunk 5]
```diff
@@ -136,6 +137,61 @@ namespace bl
...
+                    globalRef,
+                    JavaException(),
+                    BL_MSG()
+                        << "Failed to create new global reference."
+                    );
+
+                return globalRef;
+            }
+
+            void deleteGlobalRef( SAA_in const jobject object) const NOEXCEPT
```
Comment (lazar-ivanov) [inline src/include/baselib/jni/JniEnvironment.h]: space before )

File: src/include/baselib/jni/JniEnvironment.h [hunk 6]
```diff
@@ -136,6 +137,61 @@ namespace bl
...
+            {
+                m_jniEnv -> DeleteGlobalRef( object );
+            }
+
+            LocalReference< jclass > findJavaClass( SAA_in const std::string& javaClassName ) const
```
Comment (lazar-ivanov) [inline src/include/baselib/jni/JniEnvironment.h]: javaClassName -> className

File: src/include/baselib/jni/JniResourceWrappers.h [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+        template
+        <
+            typename E = void
+        >
+        class LocalReferenceDeleterT FINAL
+        {
+        public:
+
+            void operator ()( SAA_in jobject object ) const
```
Comment (lazar-ivanov) [inline src/include/baselib/jni/JniResourceWrappers.h]: SAA_in const jobject object

Comment (lazar-ivanov) [inline src/include/baselib/jni/JniResourceWrappers.h]: also you have to make this NOEXCEPT

File: src/include/baselib/jni/JniResourceWrappers.h [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+            {
+                const auto& jniEnv = JniEnvironmentT< E >::instance();
```
Comment (lazar-ivanov) [inline src/include/baselib/jni/JniResourceWrappers.h]: jniEnv -> environment

File: src/include/baselib/jni/JniResourceWrappers.h [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+        template
+        <
+            typename E = void
+        >
+        class GlobalReferenceDeleterT FINAL
+        {
+        public:
+
+            void operator ()( SAA_in jobject object ) const NOEXCEPT
```
Comment (lazar-ivanov) [inline src/include/baselib/jni/JniResourceWrappers.h]: SAA_in const jobject object

File: src/include/baselib/jni/JniResourceWrappers.h [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+            {
+                const auto& jniEnv = JniEnvironmentT< E >::instance();
```
Comment (lazar-ivanov) [inline src/include/baselib/jni/JniResourceWrappers.h]: jniEnv -> environment

Comment (lazar-ivanov) [inline src/include/baselib/jni/JniResourceWrappers.h]: 2. also probably the variable is not necessary
3. also why not just use JniEnvironment::instance()?

e.g.
JniEnvironment::instance().deleteGlobalRef( object );

File: src/include/baselib/jni/JniResourceWrappers.h (hunk 8 above)
Comment (lazar-ivanov) [inline src/include/baselib/jni/JniResourceWrappers.h]: 2. also probably the variable is not necessary
3. also why not just use JniEnvironment::instance()?

e.g.
JniEnvironment::instance().deleteLocalRef( object );

File: src/include/baselib/jni/JniResourceWrappers.h [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+            typename T
+        >
+        using GlobalReference = cpp::SafeUniquePtr< typename std::remove_pointer< T >::type, GlobalReferenceDeleterT<> >;
+
+        template
+        <
+            typename T,
+            typename E = void
+        >
+        GlobalReference< T > createGlobalReference( SAA_in const LocalReference < T >& localRef )
```
Comment (lazar-ivanov) [inline src/include/baselib/jni/JniResourceWrappers.h]: localRef -> localReference

File: src/include/baselib/jni/JniResourceWrappers.h [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+        {
+            const auto& jniEnv = JniEnvironmentT< E >::instance();
```
Comment (lazar-ivanov) [inline src/include/baselib/jni/JniResourceWrappers.h]: The variable is not necessary... see comments above, but if you want it you should remove the abbreviations (e.g. environment)

File: src/utests/utf_baselib_jni/TestJni.h [hunk 13]
```diff
@@ -79,3 +80,49 @@ UTF_AUTO_TEST_CASE( Jni_CreateJniEnvironments )
...
+
+            UTF_REQUIRE_EQUAL( jniEnv.getObjectRefType( globalRef.get() ), JNIGlobalRefType );
+        };
+
+        verifyReferences( true /* isMainThread */ );
+
+        bl::os::thread thread( verifyReferences, false /* isMainThread */ );
+        thread.join();
+    }
+}
```
Comment (lazar-ivanov) [inline src/utests/utf_baselib_jni/TestJni.h]: remove the abbreviations from the code above

File: src/include/baselib/jni/JniEnvironment.h (hunk 4 above)
Comment (alex-koren) [inline src/include/baselib/jni/JniEnvironment.h]: No error checking needed here - GetObjectRefType returns JNIInvalidRefType if the input argument is not a valid JNI object.

Comment (lazar-ivanov) [inline src/include/baselib/jni/JniEnvironment.h]: Actually it is needed IMO... passing invalid object reference indicates a bug and / or process state corruption (in which case it is best to terminate immediately)... if the returned value is JNIInvalidRefType it should RIP terminate the process. It is probably best to leave the method as NOEXCEPT and throw an exception (UnexpectedException), but wrap it in BL_NOEXCEPT_* block which should cause RIP if an exception occurs.

**Pull request closed without merging**